File: src/2818.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from math import sqrt
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def __init__(self) -> None:
        self.MODULO = 10 ** 9 + 7

    def powerOfModulo(self, base: int, exponent: int) -> int:
        retVal = 1

        # Calculate the exponentiation using binary exponentiation
        while exponent > 0:
            # If the exponent is odd, multiply the result by the base
            if exponent % 2 == 1:
                retVal = (retVal * base) % self.MODULO

            # Square the base and halve the exponent
            base = (base * base) % self.MODULO
            exponent //= 2

        return retVal

    def maximumScore(self, nums: List[int], k: int) -> int:
        retVal = 1

        numsSize = len(nums)

        primeScores = [0] * numsSize
        # Calculate the prime score for each number in nums
        for index in range(numsSize):
            num = nums[index]
            # Check for prime factors from 2 to sqrt(n)
            for factor in range(2, int(sqrt(num)) + 1):
                if num % factor != 0:
                    continue
                # Increment prime score for each prime factor
                primeScores[index] += 1
                # Remove all occurrences of the prime factor from num
                while num % factor == 0:
                    num //= factor

            # If num is still greater than or equal to 2, it's a prime factor
            if num >= 2:
                primeScores[index] += 1

        # Initialize next and previous dominant index arrays
        nextDominant = [numsSize] * numsSize
        prevDominant = [-1] * numsSize
        # Stack to store indices for monotonic decreasing prime score
        decreasingPrimeScoreStack = []
        # Calculate the next and previous dominant indices for each number
        for index in range(numsSize):
            # While the stack is not empty and the current prime score is greater than the stack's top
            while (decreasingPrimeScoreStack) and (primeScores[decreasingPrimeScoreStack[-1]] < primeScores[index]):
                topIndex = decreasingPrimeScoreStack.pop()
                # Set the next dominant element for the popped index
                nextDominant[topIndex] = index

            # If the stack is not empty, set the previous dominant element for the current index
            if decreasingPrimeScoreStack:
                prevDominant[index] = decreasingPrimeScoreStack[-1]

            # Push the current index onto the stack
            decreasingPrimeScoreStack.append(index)

        # Calculate the number of subarrays in which each element is dominant
        numOfSubarrays = [0] * numsSize
        for index in range(numsSize):
            numOfSubarrays[index] = (nextDominant[index] - index) * (index - prevDominant[index])

        # Priority queue to process elements in decreasing order of their value
        processingQueue = []
        # Push each number and its index onto the priority queue
        for index in range(numsSize):
            heappush(processingQueue, (-nums[index], index))

        # Process elements while there are operations left
        while k > 0:
            # Get the element with the maximum value from the queue
            num, index = heappop(processingQueue)
            num = -num  # Negate back to positive

            # Calculate the number of operations to apply on the current element
            operations = min(k, numOfSubarrays[index])

            # Update the score by raising the element to the power of operations
            retVal = (retVal * self.powerOfModulo(num, operations)) % self.MODULO

            # Reduce the remaining operations count
            k -= operations

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, k in zip([[8, 3, 9, 3, 8], [19, 12, 14, 6, 10, 18]], [2, 3]):
            # /* Example
            #  *  Input: nums = [8,3,9,3,8], k = 2
            #  *  Output: 81
            #  *
            #  *  Input: nums = [19,12,14,6,10,18], k = 3
            #  *  Output: 4788
            #  */
            logging.info("Input: nums = %s, k = %s", nums, k)

            retVal = pSolution.maximumScore(nums, k)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/2818.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MODULO (int)(1e9 + 7)
typedef struct {
    int value;
    int index;
} pair;
int compareStruct(const void* a, const void* b) {
    int pa = ((pair*)a)->value;
    int pb = ((pair*)b)->value;

    // ascending order
    return (pa > pb);
}
long long powerOfModulo(long long base, long long exponent) {
    long long retVal = 1;

    // Calculate the exponentiation using binary exponentiation
    while (exponent > 0) {
        // If the exponent is odd, multiply the result by the base
        if (exponent % 2 == 1) {
            retVal = ((retVal * base) % MODULO);
        }

        // Square the base and halve the exponent
        base = (base * base) % MODULO;
        exponent /= 2;
    }

    return retVal;
}
int maximumScore(int* nums, int numsSize, int k) {
    int retVal = 0;

    // Calculate the prime score for each number in nums
    int primeScores[numsSize];
    memset(primeScores, 0, sizeof(primeScores));
    for (int index = 0; index < numsSize; index++) {
        int num = nums[index];
        // Check for prime factors from 2 to sqrt(n)
        for (int factor = 2; factor <= sqrt(num); factor++) {
            if (num % factor != 0) {
                continue;
            }
            // Increment prime score for each prime factor
            primeScores[index]++;
            // Remove all occurrences of the prime factor from num
            while (num % factor == 0) {
                num /= factor;
            }
        }
        // If num is still greater than or equal to 2, it's a prime factor
        if (num >= 2) {
            primeScores[index]++;
        }
    }

    // Initialize next and previous dominant index arrays
    int nextDominant[numsSize];
    memset(nextDominant, 0, sizeof(nextDominant));
    int prevDominant[numsSize];
    memset(prevDominant, 0, sizeof(prevDominant));
    for (int i = 0; i < numsSize; i++) {
        nextDominant[i] = numsSize;
        prevDominant[i] = -1;
    }
    // Stack to store indices for monotonic decreasing prime score
    int decreasingPrimeScoreStack[numsSize];
    int decreasingPrimeScoreStackTop = -1;
    // Calculate the next and previous dominant indices for each number
    for (int index = 0; index < numsSize; index++) {
        // While the stack is not empty and the current prime score is greater than the stack's top
        while ((decreasingPrimeScoreStackTop >= 0) &&
               (primeScores[decreasingPrimeScoreStack[decreasingPrimeScoreStackTop]] < primeScores[index])) {
            int topIndex = decreasingPrimeScoreStack[decreasingPrimeScoreStackTop];
            decreasingPrimeScoreStackTop--;
            // Set the next dominant element for the popped index
            nextDominant[topIndex] = index;
        }
        // If the stack is not empty, set the previous dominant element for the current index
        if (decreasingPrimeScoreStackTop >= 0) {
            prevDominant[index] = decreasingPrimeScoreStack[decreasingPrimeScoreStackTop];
        }
        // Push the current index onto the stack
        decreasingPrimeScoreStack[++decreasingPrimeScoreStackTop] = index;
    }

    // Calculate the number of subarrays in which each element is dominant
    long long numOfSubarrays[numsSize];
    memset(numOfSubarrays, 0, sizeof(numOfSubarrays));
    for (int index = 0; index < numsSize; index++) {
        numOfSubarrays[index] = (long long)(nextDominant[index] - index) * (index - prevDominant[index]);
    }

    // Priority queue to process elements in decreasing order of their value
    pair processingQueue[numsSize];
    memset(processingQueue, 0, sizeof(processingQueue));
    int processingQueueIndex = 0;
    // Push each number and its index onto the priority queue
    for (int index = 0; index < numsSize; index++) {
        processingQueue[index].value = nums[index];
        processingQueue[index].index = index;
        processingQueueIndex++;
    }
    qsort(processingQueue, numsSize, sizeof(pair), compareStruct);

    // Process elements while there are operations left
    long long score = 1;
    while (k > 0) {
        // Get the element with the maximum value from the queue
        processingQueueIndex--;
        long long num = processingQueue[processingQueueIndex].value;
        int index = processingQueue[processingQueueIndex].index;
        // Calculate the number of operations to apply on the current element
        long long operations = fmin((long long)k, numOfSubarrays[index]);
        // Update the score by raising the element to the power of operations
        score = (score * powerOfModulo(num, operations)) % MODULO;
        // Reduce the remaining operations count
        k -= operations;
    }
    retVal = score;

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int nums[MAX_SIZE];
        int numsSize;
        int k;
    } testCase[] = {{{8, 3, 9, 3, 8}, 5, 2}, {{19, 12, 14, 6, 10, 18}, 6, 3}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: nums = [8,3,9,3,8], k = 2
     *  Output: 81
     *
     *  Input: nums = [19,12,14,6,10,18], k = 3
     *  Output: 4788
     */

    int answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: nums = [");
        for (j = 0; j < testCase[i].numsSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].nums[j]);
        }
        printf("], k = %d\n", testCase[i].k);

        answer = maximumScore(testCase[i].nums, testCase[i].numsSize, testCase[i].k);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}
